### PR TITLE
Issue #740 - System level workflow parallelism limits & priorities

### LINKED ISSUE #740: Proposal: system level workflow parallelism limits & priorities
Capturing a discussion from [the google group](https://groups.google.com/forum/#!topic/argoproj/mm8V_6u4liQ) into a GitHub issue. Read from top to bottom:

On Thursday, February 1, 2018 at 8:14:36 AM UTC-8, Rob Oxspring wrote:

I might have misunderstood but that issue seems to be primarily concerned about parallelism of steps within a workflow, which is interesting but not especially useful for our purpose as we're using very limited parallelism within a workflow anyway. 

We typically run workflows of 20-30 steps of which a few might consume large amounts of memory. We size our cluster with the expectation of a maximum of N workflows running at a time. What I'd like is to be able to enforce this limit of N active workflow resources at any time, or perhaps limit count(pods selecting label memoryhog=true)<2N, but I can't see ways of doing that. Admittedly the latter could be useful as a general Kubernetes feature request, but the former would seem reasonable as an Argo feature request?

In the absence of those feature then I guess the options would be to implement a ValidatingAdminissionWebhook, or implement queuing in front of Argo submissions. Any other obvious options I missed?

Thanks,

Rob

--------------------------------------
Hi Rob,

You are correct, the issue which I linked to is primarily concerned with limiting parallelism of *pods* within a workflow. It does not cover the limiting of *workflows* in the system at a time.

Controlling number of concurrent workflows at the controller level isn't something we've thought about, but I do think it's something that warrant's some thought and discussion. The following are the considerations in the implementation that would need to be handled:
- we would first need to introduce a Pending state at the workflow level. Pending is not actually part of the workflow state machine currently (it's an inferred state only shown in the CLI).
- If you permit Pending workflows, you now have essentially introduced a priority system in the controller, because when deciding on the next workflow to allow to run, the basic requirement would be to choose it in a FIFO fashion. The problem is that the kubernetes informers framework (of which controllers are built on top of), does not lend itself well to this mode of resource selection.
- workflow-controller is expected to eventually scale horizontally (e.g. the workflow controller might run with N replicas). When there are multiple controllers managing the same set of workflows, then the book-keeping/shared knowledge of workflow counts would be an issue. But this is a theoretical problem we can deal with at a later time.

In terms of options today, kubernetes admission webhooks might work, but I think this would not permit queuing, so it doesn't seem like a good fit. Also, I'm not sure if that mechanism supports CRDs. Your last suggestion, the external queuing on-top of argo, is probably the best option and most flexible. Admission control is something that typically wants to take a lot of other things into consideration (e.g. priority, permissions, stop-the-world, etc...). Since we view the workflow-controller as a building block for other applications, my preference is to implement the queuing at that layer. 

On a related note, Kubernetes 1.9 has an alpha feature for pod prioritization. I would want to understand the implementation of that because workflow prioritization should be implemented in the same manner.

BTW, I will file a new issue for this on github to facilitate further discussion (in case other people request the same/similar feature).

-Jesse

### PR DESCRIPTION
**Pull request by alexmt**
Closes #740

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -136,6 +137,7 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, podWorkers in
 	}
 
 	wfc.wfInformer = util.NewWorkflowInformer(wfc.restConfig, wfc.Config.Namespace, workflowResyncPeriod, wfc.tweakWorkflowlist)
+	wfc.throttler = NewThrottler(wfc.Config.Parallelism, wfc.wfQueue)
```
Comment (jessesuen) [inline workflow/controller/controller.go]: The workflow controller is supposed to notice changes to the configmap and have the settings take affect without a controller restart. I don't see anywhere where if they update the parallelism value, that the new number will take affect.

File: workflow/controller/controller.go [hunk 2]
```diff
@@ -227,12 +229,21 @@ func (wfc *WorkflowController) processNextItem() bool {
...
 		// can get here if we already added the completed=true label,
 		// but we are still draining the controller's workflow workqueue
 		return true
 	}
-	woc := newWorkflowOperationCtx(wf, wfc)
-	woc.operate()
+
+	if wfc.throttler.CanProcess(key, wf.Spec.Priority) {
+		woc := newWorkflowOperationCtx(wf, wfc)
+		woc.operate()
```
Comment (jessesuen) [inline workflow/controller/controller.go]: I think the logic should be that any workflow which has already started, should be allowed to continue to operate. So priority only takes affect during admission, but once the workflow is admitted, it should be allowed to complete. This code appears to halt all processing of any workflow (even if already running), so long as theres a higher one in the queue.

File: workflow/controller/controller.go [hunk 3]
```diff
@@ -227,12 +229,21 @@ func (wfc *WorkflowController) processNextItem() bool {
...
+		if woc.wf.Status.Completed() {
+			wfc.throttler.NotifyProcessed(key)
+		}
+	} else {
+		log.Warnf("Workflow %s processing has been postponed due to max parallelism limit", key)
```
Comment (jessesuen) [inline workflow/controller/controller.go]: Style nit: I prefer this return style instead:
```
if !wfc.throttler.CanProcess(key, wf.Spec.Priority) {
    log.Warnf("Workflow %s processing has been postponed due to max parallelism limit", key)
    return true
}

woc := newWorkflowOperationCtx(wf, wfc)
woc.operate()

```

File: workflow/controller/controller.go (hunk 2 above)
Comment (alexmt) [reply workflow/controller/controller.go]: The in-progress workflows are not halted. Throttler allows executing workflow if it is already in progress:
```
func (t *throttler) CanProcess(key interface{}, priority int) bool {
    _, isInProgress := t.processingKeys[key]
    shouldProcess := isInProgress || *t.parallelism > len(t.processingKeys)
```

Comment (jessesuen) [reply workflow/controller/controller.go]: Ok I missed that part. Thanks.

File: workflow/controller/controller.go (hunk 1 above)
Comment (alexmt) [reply workflow/controller/controller.go]: On config change controller notifies throttler using `throttler.SetParallelism` and throttler triggers required number of workflows .

```
func (t *throttler) SetParallelism(parallelism *int) {
	t.throttleLock.Lock()
	defer t.throttleLock.Unlock()
	t.parallelism = parallelism
	t.queueThrottled()
}
```

You might also check `TestSetParallelismWithPriority` unit test

File: pkg/apis/workflow/v1alpha1/types.go [hunk 4]
```diff
@@ -133,6 +133,8 @@ type WorkflowSpec struct {
 	// allowed to run before the controller terminates the workflow. A value of zero is used to
 	// terminate a Running workflow
 	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
+	// Priority is used if controller is configured to process limited number of workflows in parallel. Workflows with higher priority are processed first.
+	Priority int `json:"priority,omitempty"`
```
Comment (jessesuen) [inline pkg/apis/workflow/v1alpha1/types.go]: We should keep this consistent with the PodSpec and make this an int32 pointer:
```
Priority *int32 `json:"priority,omitempty" protobuf:"bytes,25,opt,name=priority"`
```

The difference between zero priority vs. omitted priority I think is important. For example, we may decide to support something like a default priority in the configmap, which at that point we have no way of knowing the intent.

Comment (alexmt) [reply pkg/apis/workflow/v1alpha1/types.go]: Good point. Fixed

**Pull request merged**